### PR TITLE
Fix saved filename in ModelCheckpoint if it already exists

### PR DESCRIPTION
**Pull request by rohitgr7**
## What does this PR do?

Fixes the saved filename if it's already present. It will override the file if it's already going to be deleted. Previously if `save_top_k=1` and filename is static for eg: `some_file`, it will create a file with the version for eg: filename=`some_file-v2.ckpt` if `max_epoch > 1` but it should be `some_file.ckpt` only.

The test fails on master.

## Before submitting
- [ ] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [x] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests? 
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:  

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified; _Bugfixes should be including in bug-fix release milestones (m.f.X) and features should be included in (m.X.b) releases._
 

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: tests/checkpointing/test_model_checkpoint.py [hunk 1]
```diff
@@ -986,3 +986,43 @@ def __init__(self, hparams):
...
+        limit_val_batches=2,
+        logger=None,
+        weights_summary=None,
+        progress_bar_refresh_rate=0,
+    )
+
+    model = CustomModel()
+    trainer.fit(model)
+    ckpt_files = os.listdir(tmpdir)
+    assert len(ckpt_files) == len(expected)
```
Comment (carmocca) [inline tests/checkpointing/test_model_checkpoint.py:1027]: This assert is unnecessary

File: tests/checkpointing/test_model_checkpoint.py [hunk 2]
```diff
@@ -986,3 +986,43 @@ def __init__(self, hparams):
...
+        dirpath=tmpdir,
+        filename='curr_epoch',
+        save_top_k=save_top_k,
+        monitor='epoch',
+        mode='max',
+    )
+    trainer = Trainer(
+        default_root_dir=tmpdir,
+        callbacks=[model_checkpoint],
+        max_epochs=4,
```
Comment (carmocca) [inline tests/checkpointing/test_model_checkpoint.py:1016]: can you parametrize this test for epochs (2, 3)? The person who reported the issue claimed:

> from one run to another i might get as a final model either the model with v0  or the one without?

File: pytorch_lightning/callbacks/model_checkpoint.py [hunk 3]
```diff
@@ -497,13 +495,19 @@ def _validate_monitor_key(self, trainer):
             )
             raise MisconfigurationException(m)
 
-    def _get_metric_interpolated_filepath_name(self, ckpt_name_metrics: Dict[str, Any], epoch: int, step: int):
+    def _get_metric_interpolated_filepath_name(
+        self, ckpt_name_metrics: Dict[str, Any], epoch: int, step: int, del_list: List[str] = []
```
Comment (carmocca) [inline pytorch_lightning/callbacks/model_checkpoint.py:499]: Can we rename `del_list` to something clearer like `existing_filepath`? I understand you are reusing the name in the `_update_best_and_save` method but the usage here is unrelated to the fact that it will be deleted later

File: pytorch_lightning/callbacks/model_checkpoint.py [hunk 4]
```diff
@@ -583,6 +589,8 @@ def _update_best_and_save(
         if torch.isnan(current):
             current = torch.tensor(float('inf' if self.mode == "min" else '-inf'))
 
+        filepath = self._get_metric_interpolated_filepath_name(ckpt_name_metrics, epoch, step, del_list)
```
Comment (carmocca) [inline pytorch_lightning/callbacks/model_checkpoint.py:592]: Why is `del_list` a `list`? It will always have one element if i'm not missing something.

Can you fix it?

File: pytorch_lightning/callbacks/model_checkpoint.py (hunk 3 above)
Comment (rohitgr7) [reply pytorch_lightning/callbacks/model_checkpoint.py:499]: changed it to del_filepath. I believe it self explains that this file will be deleted so no need to update.

File: tests/checkpointing/test_model_checkpoint.py (hunk 1 above)
Comment (rohitgr7) [reply tests/checkpointing/test_model_checkpoint.py:1027]: True, updated.

File: pytorch_lightning/callbacks/model_checkpoint.py (hunk 4 above)
Comment (rohitgr7) [reply pytorch_lightning/callbacks/model_checkpoint.py:592]: Great catch! Updated.

File: tests/checkpointing/test_model_checkpoint.py (hunk 2 above)
Comment (rohitgr7) [reply tests/checkpointing/test_model_checkpoint.py]: done.

File: pytorch_lightning/callbacks/model_checkpoint.py [hunk 5]
```diff
@@ -497,13 +495,23 @@ def _validate_monitor_key(self, trainer):
...
+        step: int,
+        del_filepath: Optional[str] = None
+    ) -> str:
         filepath = self.format_checkpoint_name(epoch, step, ckpt_name_metrics)
+
         version_cnt = 0
         while self._fs.exists(filepath):
+            if filepath == del_filepath:
+                break
+
```
Comment (carmocca) [inline pytorch_lightning/callbacks/model_checkpoint.py:511]:
```suggestion
        while self._fs.exists(filepath) and filepath != del_filepath:
```

**Review by carmocca — APPROVED** [review]

**Comment by rohitgr7** [comment]
Thanks @carmocca for the review :)

**Comment by carmocca** [comment]
Unrelated to the PR but, is there a reason for version starting at 0?

```
file.ckpt  # I find this confusing, kind of implies this is -v-1
file-v0.ckpt
file-v1.ckpt
...
```

instead of:

```
file.ckpt
file-v1.ckpt
file-v2.ckpt
...
```

**Comment by Borda** [comment]
> Unrelated to the PR but, is there a reason for version starting at 0?

there is no special reason :]
cc: @PyTorchLightning/core-contributors

**Comment by rohitgr7** [comment]
> file.ckpt  # I find this confusing, kind of implies this is -v-1
file-v0.ckpt
file-v1.ckpt
...

no strong opinion here. To me, it looks like `file.ckpt` the filename user requested, and `file-v0.ckpt, ...` files created with duplicate names starting with 0 version index.

**Comment by carmocca** [comment]
Imagine you know nothing about Lightning and you are given a copy of these checkpoints so you can run a model on some data. Would it be clear which is the first checkpoint?

**Comment by rohitgr7** [comment]
> Would it be clear which is the first checkpoint?

monitor values = [10, 20, 15, 5], mode='min'
`file.ckpt` might not always be the first checkpoint.

File: tests/checkpointing/test_model_checkpoint.py [hunk 6]
```diff
@@ -986,3 +986,43 @@ def __init__(self, hparams):
...
         # make sure it's not AttributeDict
         assert type(ckpt[model.CHECKPOINT_HYPER_PARAMS_KEY]) == hparams_type
+
+
+@pytest.mark.parametrize('max_epochs', [3, 4])
+@pytest.mark.parametrize(
+    'save_top_k, expected',
+    [
+        (1, ['curr_epoch.ckpt']),
+        (2, ['curr_epoch.ckpt', 'curr_epoch-v0.ckpt']),
```
Comment (Borda) [inline tests/checkpointing/test_model_checkpoint.py:996]: I am thinking if in such case we shall rename the previous one
```suggestion
        (2, ['curr_epoch.ckpt-v0', 'curr_epoch-v1.ckpt']),
```
cc: @PyTorchLightning/core-contributors

Comment (rohitgr7) [reply tests/checkpointing/test_model_checkpoint.py:996]: I like it. Just thinking about how to do it the right way.

Comment (carmocca) [reply tests/checkpointing/test_model_checkpoint.py:996]: I also like the idea, solves what I commented here https://github.com/PyTorchLightning/pytorch-lightning/pull/4861#issuecomment-734458571. But better to do it in another PR, right?

Comment (Borda) [reply tests/checkpointing/test_model_checkpoint.py:996]: basically when you find existing and your internal counter is the min (you run counter from 1 not 0) then the fist you move to v0, it shall not be so hard, just not sure if fssspec allows it do simply...

Comment (rohitgr7) [reply tests/checkpointing/test_model_checkpoint.py:996]: thanks, will try to get this done today.

File: tests/checkpointing/test_model_checkpoint.py [hunk 7]
```diff
@@ -1026,3 +1026,6 @@ def validation_epoch_end(self, outputs):
     trainer.fit(model)
     ckpt_files = os.listdir(tmpdir)
     assert set(ckpt_files) == set(expected)
+
+    expected_epoch_in_files = sorted([pl_load(os.path.join(tmpdir, f))['epoch'] - 1 for f in ckpt_files])
+    assert expected_epoch_in_files == sorted(list(range(max_epochs))[-save_top_k:])
```
Comment (carmocca) [inline tests/checkpointing/test_model_checkpoint.py:1042]:
```suggestion
    assert expected_epoch_in_files == list(range(max_epochs - save_top_k, max_epochs))
```

File: pytorch_lightning/callbacks/model_checkpoint.py [hunk 8]
```diff
@@ -504,8 +504,24 @@ def _get_metric_interpolated_filepath_name(
     ) -> str:
         filepath = self.format_checkpoint_name(epoch, step, ckpt_name_metrics)
 
-        version_cnt = 0
-        while self._fs.exists(filepath) and filepath != del_filepath:
+        version_cnt = 1
+        old_ckpt_ver_0 = self.format_checkpoint_name(epoch, step, ckpt_name_metrics, ver=0)
```
Comment (carmocca) [inline pytorch_lightning/callbacks/model_checkpoint.py:508]: does this work if the checkpoint name contains a quantity (e.g. val loss)? wouldn't `ckpt_name_metrics` contain different values each time?

**Comment by rohitgr7** [comment]
let's do https://github.com/PyTorchLightning/pytorch-lightning/pull/4861#discussion_r531787757 in another PR, since we need to make sure to update the states in ModelCheckpoint when we rename the file. Reverting the changes.

**Comment by rohitgr7** [comment]
@carmocca wanna give this a shot? https://github.com/PyTorchLightning/pytorch-lightning/pull/4861#discussion_r531787757

File: tests/checkpointing/test_model_checkpoint.py (hunk 7 above)
Comment (rohitgr7) [reply tests/checkpointing/test_model_checkpoint.py]: thanks

**Comment by carmocca** [comment]
> @carmocca wanna give this a shot? [#4861 (comment)](https://github.com/PyTorchLightning/pytorch-lightning/pull/4861#discussion_r531787757)

Sure, I will. Most likely this weekend

**Review by awaelchli — APPROVED** [review]

File: tests/checkpointing/test_model_checkpoint.py (hunk 6 above)
Comment (Borda) [reply tests/checkpointing/test_model_checkpoint.py:1030]: is this implemented?

Comment (Borda) [reply tests/checkpointing/test_model_checkpoint.py:1030]: ok I see #5000

Comment (rohitgr7) [reply tests/checkpointing/test_model_checkpoint.py:1030]: it's wip in a separate PR. https://github.com/PyTorchLightning/pytorch-lightning/issues/5000

File: tests/checkpointing/test_model_checkpoint.py [hunk 9]
```diff
@@ -1020,3 +1020,46 @@ def __init__(self, hparams):
...
+        (2, ['curr_epoch.ckpt', 'curr_epoch-v0.ckpt']),
+    ]
+)
+def test_model_checkpoint_file_already_exists(tmpdir, max_epochs, save_top_k, expected):
+    """
+    Test that version is added to filename if required and it already exists in dirpath.
+    """
+    class CustomModel(LogInTwoMethods):
+        def validation_epoch_end(self, outputs):
+            return {'epoch': self.current_epoch}
```
Comment (carmocca) [inline tests/checkpointing/test_model_checkpoint.py:1039]: You don't need to subclass `BoringModel` to monitor `epoch`. You can do it already with `BoringModel()`. See these tests for example:

https://github.com/PyTorchLightning/pytorch-lightning/pull/5008/files#diff-44594fac97ae772bb165301c22fdc1b25cd481589502dd7413c6953289cbb718R1033-R1044

File: tests/checkpointing/test_model_checkpoint.py (hunk 6 above)
Comment (carmocca) [reply tests/checkpointing/test_model_checkpoint.py:1030]: The PR is https://github.com/PyTorchLightning/pytorch-lightning/pull/5008 (not WIP)

File: tests/checkpointing/test_model_checkpoint.py (hunk 9 above)
Comment (rohitgr7) [reply tests/checkpointing/test_model_checkpoint.py]: yeah, I was actually monitoring `val_acc` initially.

**Review by s-rog — APPROVED** [review]
